src/main/kotlin/sliv/tool/scene/view/LandmarkView.kt

```diff
@@ -1,27 +1,153 @@
             return when (landmark) {
-                is Landmark.Keypoint -> KeypointView(landmark, scale)
-                is Landmark.Line -> LineView(landmark, scale)
-                is Landmark.Plane -> PlaneView(landmark, scale)
+                is Landmark.Keypoint -> KeypointView(landmark, scale, frameTimestamp, stateSynchronizationManager)
+                is Landmark.Line -> LineView(landmark, scale, frameTimestamp, stateSynchronizationManager)
+                is Landmark.Plane -> PlaneView(landmark, scale, frameTimestamp, stateSynchronizationManager)
             }
         }
     }
 
-    abstract val shape: Shape
+    // When shape is created in an inheritor
+    // setUpShape() should be called to set up common features for all landmarks
+    abstract val node: Node
+
+    private var state: LandmarkState = LandmarkState.Ordinary
+
     var scale: Double = scale
         set(value) {
             field = value
             scaleChanged()
         }
 
+    private val landmarkSelectedEventHandler: (LandmarkEventArgs) -> Unit = { eventArgs ->
+        if (frameTimestamp != eventArgs.frameTimestamp && eventArgs.uid == landmark.uid) {
+            select()
+        }
+    }
+
+    private val landmarkUnselectedEventHandler: (LandmarkEventArgs) -> Unit = { eventArgs ->
+        if (frameTimestamp != eventArgs.frameTimestamp && eventArgs.uid == landmark.uid) {
+            unselect()
+        }
+    }
+
+    private val landmarkMouseEnteredEventHandler: (LandmarkEventArgs) -> Unit = { eventArgs ->
+        if (frameTimestamp != eventArgs.frameTimestamp && eventArgs.uid == landmark.uid) {
+            handleMouseEntered()
+        }
+    }
+
+    private val landmarkMouseExitedEventHandler: (LandmarkEventArgs) -> Unit = { eventArgs ->
+        if (frameTimestamp != eventArgs.frameTimestamp && eventArgs.uid == landmark.uid) {
+            handleMouseExited()
+        }
+    }
+
+    init {
+        stateSynchronizationManager.landmarkSelected += landmarkSelectedEventHandler
+        stateSynchronizationManager.landmarkUnselected += landmarkUnselectedEventHandler
+        stateSynchronizationManager.landmarkMouseEntered += landmarkMouseEnteredEventHandler
+        stateSynchronizationManager.landmarkMouseExited += landmarkMouseExitedEventHandler
+    }
+
+    fun dispose() {
+        stateSynchronizationManager.landmarkSelected -= landmarkSelectedEventHandler
+        stateSynchronizationManager.landmarkUnselected -= landmarkUnselectedEventHandler
+        stateSynchronizationManager.landmarkMouseEntered -= landmarkMouseEnteredEventHandler
+        stateSynchronizationManager.landmarkMouseExited -= landmarkMouseExitedEventHandler
+    }
+
+    // Set up common shape properties
+    // Can not be called during LandmarkView initialization because shape is created by inheritors
+    protected fun setUpShape(shape: Shape) {
```

Comment:
use abstract fun to init shape and separate casted variable in inheritors, don't think that this method will be called by them always

Reply:
why? you will just init shapes there, common logic is still here. only driver code changes